src/modules/service.js

```diff
@@ -1,94 +1,62 @@
-              // returns ugly HTML in the response body.
-              body = 'Not found.';
-            }
-            fulfill(this._buildResult(response, body));
-          } else {
-            reject(err);
-          }
-        });
-      } else if (this.config.cacheMode === 'dryrun') {
-        this.log.debug(`dryrun for ${options.uri}`);
-        let body = this.cache.read(options.uriCache);
-        if (body) {
-          let response = {};
-          response.statusCode = 200;
-          fulfill(this._buildResult(response, body));
-        } else {
-          request.get(options, (err, response, body) => {
-            if (!err) {
-              if (response.statusCode === 404) {
-                // special case for 404 because the Student Web Service
-                // returns ugly HTML in the response body.
-                body = 'Not found.';
-              }
-              fulfill(this._buildResult(response, body));
-            } else {
-              reject(err);
-            }
-          });
-        }
-      } else if (this.config.cacheMode === 'record') {
-        this.log.debug(`record -- ${options.uri}`);
-        let body = this.cache.read(options.uriCache);
-        if (body) {
-          let response = {};
-          response.statusCode = 200;
-          fulfill(this._buildResult(response, body));
-        } else {
-          request.get(options, (err, response, body) => {
-            if (!err) {
-              if (response.statusCode === 200) {
-                this.cache.write(options.uriCache, body, true);
-              } else if (response.statusCode === 404) {
-                // special case for 404 because the Student Web Service
-                // returns ugly HTML in the response body.
-                body = 'Not found.';
-              }
-              fulfill(this._buildResult(response, body));
-            } else {
-              reject(err);
-            }
-          });
-        }
+    if (cacheMode !== 'wild') {
+      const body = this.cache.read(options.uriCache);
+
+      if (body) {
+        response.statusCode = 200;
+        response.body = body;
+        this.log.debug(`${cacheMode} cache hit for ${options.url}`);
+      } else {
+        this.log.debug(`${cacheMode} cache miss for ${options.url}`);
       }
-    });
+    }
+
+
+    response = await got.get(options)
+      .catch((err) => err.response);
+
+    this.log.debug(`GET -- ${options.url}`);
+
+    if (cacheMode === 'record' && !!response.body) {
+      this.cache.write(options.uriCache, response.body, true);
+    }
+
+    return this.buildResult(response);
   }
 
-  _buildResult(response, body) {
-    let result = {};
+  buildResult(response) {
+    const { body } = response;
+    this.log.trace(`Response body: ${body}`);
+    const result = {};
     result.statusCode = response.statusCode;
     if (response.statusCode !== 200) {
-      if (!this._isJson(body)) {
+      if (!Service.isJson(body)) {
```

Comment:
Maybe ternary on this one since many if else around here?
`result.message = !Service.isJson(body) ? body  : JSON.parse(body);`
just for the readability?

Reply:
While I like to avoid ternary because it's often hard to figure out what's going on, this is a good suggestion to clarify the code. Thanks!